documentation/api/addAssertion.md

```diff
@@ -10,82 +10,237 @@ expect.addAssertion(pattern, handler);
-or one of the parameters to the assertion. In this case the assertion
-is only defined for arrays. If no subject type is specified,
-the assertion will be defined for the type `any`, and would
-be applicable any type. See the `Extending Unexpected with new types`
-section for more information about the type system in Unexpected.
-
-A word in square brackets represents a flag that can either be
-there or not. If the flag is present `expect.flags[flag]` will contain
-the value `true`. In this case `not` is a flag. When a flag it present
-in a nested `expect` it will be inserted if the flag is present;
-otherwise it will be removed. Text that is in parentheses with
-vertical bars between them are treated as alternative texts that can
-be used. In this case you can write _ordered_ as an alternative to
-_sorted_.
+```js
+expect([1, 2, 3], 'to have item', 2);
+expect([1, 2, 3], 'not to have item', 4);
+```
 
-An assertion can only have a single assertion string, which must be
-provided after the subject type. This means that you cannot have
-more words, flags, and alternations after the first type.
+Flags are made available to the handler function as an `expect.flags` object,
+where the keys are the names of the flags and the values are `true`, if the flag
+is used, or otherwise `undefined`.
 
-The second and last parameter to `addAssertion` is function that will
-be called when `expect` is invoked with an expectation matching the
-type and pattern of the assertion.
+This example could be improved further. Since
+[to contain](../../assertions/array-like/to-contain/) also supports the `not`
+flag, one can propagate the flag to that assertion as follows:
 
-So in this case, when `expect` is called the following way:
+<!-- unexpected-markdown freshExpect:true -->
 
-<!-- unexpected-markdown evaluate:false -->
+```js
+expect.addAssertion('<array> [not] to have item <any>', function(
+  expect,
+  subject,
+  value
+) {
+  expect(subject, '[not] to contain', value);
+});
+```
+
+In this way, when `to have item` is invoked with the `not` flag, that flag will
+be passed along to `to contain`.
+
+When flags are propagated, one can also invert the flag as follows:
+
+<!-- unexpected-markdown freshExpect:true -->
 
 ```js
-expect([3, 2, 1], 'to be sorted', reverse);
+expect.addAssertion('<array> [not] to have item <any>', function(
+  expect,
+  subject,
+  value
+) {
+  expect(subject, '[!not] to contain', value);
+});
 ```
 
-The handler to our assertion will be called with the values the
-following way, where the _not_ flag in the nested expect will be
-removed:
+This means that if `to have item` is invoked with the `not` flag, that flag will
+not be propagated to `to contain` - and vice versa:
 
-<!-- unexpected-markdown evaluate:false -->
-<!-- eslint-skip -->
+```js
+expect([1, 2, 3], 'not to have item', 2);
+expect([1, 2, 3], 'to have item', 4);
+```
+
+Fun with flags, right? Flags can also be used to define optional filler words
```

Comment:
It can't be done with alternations. For alternations, at least one of them must be present in the invocation, therefore it's not really a filler word.

I can remove this sentence if you'd like.

Reply:
@joelmukuthu hmmm I might be wrong, but I'm pretty sure you are allowed to have an empty alternation - ah, remembered where I did: https://github.com/alexjeffburke/unexpected-events/blob/master/lib/unexpectedEvents.js#L206

Reply:
We also expand all the combinations of the alternations and also populate `expect.alternations` for the assertion, so the cost is the same. Personally I think `<EventEmitter> for the [first] event [values] on <string> <assertion>` reads better.

Reply:
@papandreou - ah okay. In that light I’m happy to withdraw the suggestion :)